zerver/lib/email_mirror_server.py

```diff
@@ -0,0 +1,190 @@
+)
+from zerver.lib.email_mirror_helpers import (
+    ZulipEmailForwardError,
+    get_email_gateway_message_string_from_address,
+)
+from zerver.lib.exceptions import RateLimitedError
+from zerver.lib.logging_util import log_to_file
+from zerver.lib.queue import queue_json_publish_rollback_unsafe
+
+logger = logging.getLogger("zerver.lib.email_mirror")
+log_to_file(logger, settings.EMAIL_MIRROR_LOG_PATH)
+
+
+def send_to_postmaster(msg: email.message.Message) -> None:
+    # RFC5321 says:
+    #   Any system that includes an SMTP server supporting mail relaying or
+    #   delivery MUST support the reserved mailbox "postmaster" as a case-
+    #   insensitive local name.  This postmaster address is not strictly
+    #   necessary if the server always returns 554 on connection opening (as
+    #   described in Section 3.1).  The requirement to accept mail for
+    #   postmaster implies that RCPT commands that specify a mailbox for
+    #   postmaster at any of the domains for which the SMTP server provides
+    #   mail service, as well as the special case of "RCPT TO:<Postmaster>"
+    #   (with no domain specification), MUST be supported.
+    #
+    # We forward such mail to the ZULIP_ADMINISTRATOR.
+    mail = DjangoEmailMultiAlternatives(
+        subject=f"Mail to postmaster: {msg['Subject']}",
+        from_email=settings.NOREPLY_EMAIL_ADDRESS,
+        to=[settings.ZULIP_ADMINISTRATOR],
+    )
+    mail.attach(None, msg, "message/rfc822")
+    try:
+        mail.send()
+    except smtplib.SMTPResponseException as e:
+        logger.exception(
+            "Error sending bounce email to %s with error code %s: %s",
+            mail.to,
+            e.smtp_code,
+            e.smtp_error,
+            stack_info=True,
+        )
+    except smtplib.SMTPException as e:
+        logger.exception("Error sending bounce email to %s: %s", mail.to, str(e), stack_info=True)
+
+
+class ZulipMessageHandler(MessageHandler):
+    def __init__(self) -> None:
+        super().__init__(email.message.Message)
+
+    async def handle_RCPT(
+        self,
+        server: SMTP,
+        session: Session,
+        envelope: Envelope,
+        address: str,
+        rcpt_options: list[str],
+    ) -> str:
+        # Rewrite all postmaster email addresses to just "postmaster"
+        if address.lower() == "postmaster":
+            envelope.rcpt_tos.append("postmaster")
+            return "250 Continue"
+        try:
+            if get_email_gateway_message_string_from_address(address).lower() == "postmaster":
+                envelope.rcpt_tos.append("postmaster")
+                return "250 Continue"
+        except ZulipEmailForwardError:
+            pass
+
+        try:
+            await sync_to_async(validate_to_address)(address)
+            if not is_missed_message_address(address):
+                # Missed message addresses are one-time use, so we don't need
+                # to worry about emails to them resulting in message spam.
```

Comment:
Is this still true? I feel like we may have removed the single-use limit on those...

Reply:
Good catch -- 7efd59b6d7d7f8b98c69086fa7cda7a2e0179091.

Reply:
Do we want to impose a rate-limit here?  That'd be new.  But probably also very safe.

Reply:
I don't think it's necessary.